src/leveled_sst.erl

```diff
@@ -1725,20 +1764,100 @@ hmac(Int) when is_integer(Int) ->
 %%
 %% This implementation of the SlotIndex uses leveled_tree
 
-from_list(SlotList) ->
-    leveled_tree:from_orderedlist(SlotList, ?TREE_TYPE, ?TREE_SIZE).
+from_list(SlotList, FirstKey, LastKey) ->
+    FilterFun = get_filterfun(FirstKey, LastKey),
+    FilteredList =
+        lists:map(fun({K, S}) -> {FilterFun(K), S} end, SlotList),
+    {leveled_tree:from_orderedlist(FilteredList, ?TREE_TYPE, ?TREE_SIZE),
+        FilterFun}.
+
+-spec get_filterfun(
+    leveled_codec:ledger_key(), leveled_codec:ledger_key()) ->
+        fun((leveled_codec:ledger_key())
+            -> leveled_codec:ledger_key()|leveled_codec:slimmed_key()).
+get_filterfun(
+        {Tag, Bucket, {Field, FT}, FK}, {Tag, Bucket, {Field, LT}, LK})
+            when is_binary(FT), is_binary(FK), is_binary(LT), is_binary(LK) ->
+    case {binary:longest_common_prefix([FT, LT]), byte_size(FT)} of
+        {N, M} when N > 0, M > N ->
+            <<Prefix:N/binary, _Rest/binary>> = FT,
+            term_prefix_filter(N, Prefix);
+        _ ->
+            fun term_filter/1
+    end;
+get_filterfun(
+        {Tag, Bucket, FK, null}, {Tag, Bucket, LK, null})
+            when is_binary(FK), is_binary(LK), FK < LK ->
+    case {binary:longest_common_prefix([FK, LK]), byte_size(FK)} of
+        {N, M} when N > 0, M > N ->
+            <<Prefix:N/binary, _Rest/binary>> = FK,
+            key_prefix_filter(N, Prefix);
+        _ ->
+            fun key_filter/1
+        
+    end;
+get_filterfun(_FirstKey, _LastKey) ->
+    fun null_filter/1.
+
+-spec null_filter(leveled_codec:ledger_key()) -> leveled_codec:ledger_key().
+null_filter(Key) -> Key.
+
+-spec key_filter(leveled_codec:ledger_key()) -> leveled_codec:slimmed_key().
+key_filter({_Tag, _Bucket, Key, null}) -> Key.
+
+-spec term_filter(leveled_codec:ledger_key()) -> leveled_codec:slimmed_key().
+term_filter({_Tag, _Bucket, {_Field, Term}, Key}) -> {Term, Key}.
+
+-spec key_prefix_filter(
+    pos_integer(), binary()) ->
+        fun((leveled_codec:ledger_key()) -> leveled_codec:slimmed_key()).
+key_prefix_filter(N, Prefix) ->
+    fun({_Tag, _Bucket, Key, null}) ->
+        case Key of
+            null ->
+                null;
+            K when byte_size(K) =< N ->
+                null;
+            <<QueryKey:N/binary, Suffix/binary>> ->
+                case QueryKey of
+                    Prefix ->
+                        Suffix;
+                    _ ->
+                        null
+                end
+        end
+    end.
 
-lookup_slot(Key, Tree) ->
+-spec term_prefix_filter(
+    pos_integer(), binary()) ->
+        fun((leveled_codec:ledger_key()) -> leveled_codec:slimmed_key()).
+term_prefix_filter(N, Prefix) ->
+    fun({_Tag, _Bucket, {_Field, Term}, Key}) ->
+        case Term of
+            T when byte_size(T) =< N ->
+                null;
+            <<QueryKey:N/binary, Suffix/binary>> ->
+                case QueryKey of
```

Comment:
Should this be QueryTerm as opposed to QueryKey for clarity?

Reply:
I think the interim assignment to anything is actually unnecessary.  We actually only have two cases, where it maps to <<Prefix:N/binary, Suffix/binary>> ... or where the result should be null.  This is the same in the key_prefix_filter/2 and the term_prefix_filter/2.

Reply:
https://github.com/martinsumner/leveled/pull/383/commits/76ed44dea489d1fd009d9d6123d92199deb7a72a

Reply:
I think this change allows `Suffix` to be null (when Key == Prefix), which wasn't allowed previously due to `K when byte_size(K) =< N`
I don't think this will cause a problem though as the return Suffix would be null which is what was being returned previously
Just wanted to point it out in case you knew of any unintended side effect here

Reply:
In this case the Suffix would be <<>> and not null I think.  However, this is a good point.  Let me think